cmd/soroban-cli/src/commands/contract/build.rs

```diff
@@ -229,3 +232,81 @@ impl Cmd {
         cmd.exec()
     }
 }
+
+/// Configure cargo/rustc to replace absolute paths in panic messages / debuginfo
+/// with relative paths.
+///
+/// This is required for reproducible builds.
+///
+/// This works for paths to crates in the registry. The compiler already does
+/// something similar for standard library paths and local paths. It may not
+/// work for crates that come from other sources, including the standard library
+/// compiled from source, though it may be possible to accomodate such cases in
+/// the future.
+///
+/// This in theory breaks the ability of debuggers to find source code, but
+/// since we are only targetting wasm, which is not typically run in a debugger,
+/// and stellar-cli only compiles contracts in release mode, the impact is on
+/// debugging is expected to be minimal.
+///
+/// This works by setting the `CARGO_ENCODED_RUSTFLAGS` environment variable,
+/// with appropriate `--remap-path-prefix` option. It preserves the values of an
+/// existing `CARGO_ENCODED_RUSTFLAGS` (or `RUSTFLAGS`).
+///
+/// This must be done via `RUSTFLAGS` and not as arguments to `cargo rustc`
+/// because the latter only applies to the crate directly being compiled, while
+/// `RUSTFLAGS` applies to all crates, including dependencies.
+///
+/// Note that a consequence of setting `RUSTFLAGS` via the environment is that
+/// cargo will ignore any `build.rustflags`, etc. configuration values on
```

Comment:
Do you know if we passed the values via `CARGO_BUILD_RUSTFLAGS`, or via `CARGO_TARGET_wasm32-unknown-unknown_RUSTFLAGS`, would they override or append / be extra? I saw these env vars listed here:
https://doc.rust-lang.org/cargo/reference/environment-variables.html#configuration-environment-variables

Reply:
All variants of rustflags are mutually exclusive - one will "win" and the others will be ignored (cc https://doc.rust-lang.org/cargo/reference/config.html#buildrustflags). `CARGO_ENCODED_RUSTFLAGS` is the highest priority, so our use here will ensure we get the results we want while possibly ignoring what the user wants per their `build.rustflags` config file entry.

Using the target-specific rustflags is an interesting idea though since we could avoid messing with build scripts - but then somebody could override our rustflags with the non-target-specific rustflags env var (which might not be so bad). I don't think there's much risk in messing with the `file!` macro in build scripts though.

Reply:
From a re-read of cargo's rustflags handling, rustflags may not apply to build scripts at all anyway.

Reply:
FWIW I did say that "all versions are mutually exclusive" though I do see signs that cargo merges config values from its hierarchy of config files, so if I set `build.rustflags` in `~/.cargo/config.toml` and also in ~/myproject/.cargo/config.toml` it may be merging them; but it definitely doesn't merge `RUSTFLAGS` with `CARGO_ENCODED_RUSTFLAGS` etc.

Reply:
> Using the target-specific rustflags is an interesting idea though since we could avoid messing with build scripts - but then somebody could override our rustflags with the non-target-specific rustflags env var (which might not be so bad).

I just re-read https://doc.rust-lang.org/cargo/reference/config.html#buildrustflags and realise that target specific rustflags actually take precedence over the build.rustflags, so I guess setting target specific wouldn't necessarily be better at allowing them to override.

`build.rustflags` seems to be the last one considered, so it should probably be the one we use, which I think we can set with `CARGO_BUILD_RUSTFLAGS`?